modules/apps/frontend-js/frontend-js-web/src/main/java/com/liferay/frontend/js/web/internal/servlet/taglib/aui/PortletDataRendererImpl.java

```diff
@@ -0,0 +1,168 @@
+/**
+ * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
+ *
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License as published by the Free
+ * Software Foundation; either version 2.1 of the License, or (at your option)
+ * any later version.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
+ * details.
+ */
+
+package com.liferay.frontend.js.web.internal.servlet.taglib.aui;
+
+import com.liferay.petra.string.StringPool;
+import com.liferay.portal.kernel.servlet.taglib.aui.AMDRequire;
+import com.liferay.portal.kernel.servlet.taglib.aui.ESMImport;
+import com.liferay.portal.kernel.servlet.taglib.aui.PortletData;
+import com.liferay.portal.kernel.servlet.taglib.aui.PortletDataRenderer;
+
+import java.io.CharArrayWriter;
+import java.io.IOException;
+import java.io.Writer;
+
+import java.util.Collection;
+
+import org.osgi.service.component.annotations.Component;
+
+/**
+ * @author Iván Zaera Avellón
+ */
+@Component(immediate = true, service = PortletDataRenderer.class)
+public class PortletDataRendererImpl implements PortletDataRenderer {
+
+	@Override
+	public void writeTo(Writer writer, Collection<PortletData> portletDatas)
+		throws IOException {
+
+		PortletDataCollection portletDataCollection = new PortletDataCollection(
+			portletDatas);
+
+		if (portletDataCollection.isUseESM()) {
+			writer.write("<script type=\"module\">\n// <![CDATA[\n");
+		}
+		else {
+			writer.write("<script type=\"text/javascript\">\n// <![CDATA[\n");
+		}
+
+		Collection<ESMImport> esmImports =
+			portletDataCollection.getESMImports();
+
+		if (!esmImports.isEmpty()) {
+			_writeESMProlog(writer, esmImports);
+		}
+
+		if (!portletDataCollection.isUseESM()) {
+			writer.write("(function() {\n");
+		}
+
+		writer.write("var $ = AUI.$;var _ = AUI._;\n");
+
+		Collection<AMDRequire> amdRequires =
+			portletDataCollection.getAMDRequires();
+
+		if (!amdRequires.isEmpty()) {
+			_writeAMDProlog(writer, amdRequires);
+		}
+
+		Collection<String> auiModules = portletDataCollection.getAUIModules();
+
+		if (!auiModules.isEmpty()) {
+			_writeAUIProlog(writer, auiModules);
+		}
+
+		writer.write(portletDataCollection.getContent());
+
+		if (!auiModules.isEmpty()) {
+			_writeAUIEpilog(writer);
+		}
+
+		if (!amdRequires.isEmpty()) {
+			_writeAMDEpilog(writer);
+		}
+
+		if (!portletDataCollection.isUseESM()) {
+			writer.write("})();\n");
+		}
+
+		writer.write("\n// ]]>\n</script>");
```

Comment:
Are all these `\n` necessary for it to work or is this just for readability in the web source?